keps/sig-storage/2639-secret-configmap-protection/README.md

```diff
@@ -0,0 +1,634 @@
+# KEP-2639: Secret/ConfigMap Protection
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+    - [Story 3](#story-3)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
+    - [Beta -&gt; GA Graduation](#beta---ga-graduation)
+    - [Removing a Deprecated Flag](#removing-a-deprecated-flag)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a feature to protect Secrets/ConfigMaps while it is in use. Currently, users can delete a secret that is being used by other resources, like Pods, PVs, and VolumeSnapshots. This may have negative impact on the resouces using the secret and it may result in data loss.
```

Comment:
Xref https://github.com/kubernetes/enhancements/pull/2840

Reply:
@wojtek-t

Thank you for your review.

Actually, it is referenced in the first line of the "Design Details" section, but I agree that it would be good to make a reference in the very beginning. Any good ways to reference here?

Would it make sense to just write `Xref #2840` here?


In addition, can/should I also change PRR reviewer for this KEP to you?
(This KEP and #2840 are very related, so it may be good to be reviewed by the same person).

Reply:
> In addition, can/should I also change PRR reviewer for this KEP to you?
(This KEP and #2840 are very related, so it may be good to be reviewed by the same person).

I need to dig into this one deeper, but I'm not convinced from the very high level glance, why we need both. Wouldn't #2840 solve these problems too?

> In addition, can/should I also change PRR reviewer for this KEP to you?

Yes - please do

Reply:
> I need to dig into this one deeper, but I'm not convinced from the very high level glance, why we need both. Wouldn't #2840 solve these problems too?

Actually - it seems that I just glanced too quickly.

My concern was that I would like to avoid introducing a new API here and instead rely on the API introduced in that other KEP. It seems that it's exactly what it is proposing (or maybe it wasn't initially and this changed in the meantime?)

Reply:
#2840 provides Lien API to protect resources in general, and this KEP only handles protection for Secret/ConfigMap by using special knowledge on when we should regard a secret as in-use. So, the only works for the controllers in this KEP are (1) watching Pods, PVCs, Snapshots, and so on and (2) adding liens to Secrets/Configmaps according to the watched resources.

Reply:
Added #2840 as a non-goal for this KEP.

Reply:
> In addition, can/should I also change PRR reviewer for this KEP to you?
> 
> Yes - please do

Also, updated the PRR reviewer. Thank you! @wojtek-t